### PR TITLE
v2: added WebhooksResource

### PR DESCRIPTION
**Pull request by oxtoacart**
Updates tailscale/corp#21867

### CODE REVIEW

File: v2/webhooks.go [hunk 1]
```diff
@@ -0,0 +1,152 @@
+package tailscale
```
Comment (oxtoacart) [inline v2/webhooks.go:1]: This file copied almost verbatim from V1. I'll point out differences in PR.

File: v2/webhooks.go [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+
+	// CreateWebhookRequest type describes the configuration for creating a Webhook.
+	CreateWebhookRequest struct {
+		EndpointURL   string                    `json:"endpointUrl"`
+		ProviderType  WebhookProviderType       `json:"providerType"`
+		Subscriptions []WebhookSubscriptionType `json:"subscriptions"`
+	}
+)
+
+type WebhooksResource struct {
```
Comment (oxtoacart) [inline v2/webhooks.go:64]: New - organizes functions into a Resource.

File: v2/webhooks.go [hunk 3]
```diff
@@ -0,0 +1,152 @@
...
+	*Client
+}
+
+// Create creates a new webhook with the specifications provided in the CreateWebhookRequest.
+// Returns a Webhook if successful.
+func (c *WebhooksResource) Create(ctx context.Context, request CreateWebhookRequest) (*Webhook, error) {
+	req, err := c.buildRequest(ctx, http.MethodPost, c.buildTailnetURL("webhooks"), requestBody(request))
```
Comment (oxtoacart) [inline v2/webhooks.go:71]: URLs are built using the new `buildTailnetURL` and `buildURL` functions.

File: v2/webhooks.go [hunk 4]
```diff
@@ -0,0 +1,152 @@
...
+	}
+)
+
+type WebhooksResource struct {
+	*Client
+}
+
+// Create creates a new webhook with the specifications provided in the CreateWebhookRequest.
+// Returns a Webhook if successful.
+func (c *WebhooksResource) Create(ctx context.Context, request CreateWebhookRequest) (*Webhook, error) {
```
Comment (oxtoacart) [inline v2/webhooks.go:70]: `Webhook` has been removed from function names since it's already named in the Resource.

File: v2/webhooks_test.go [hunk 5]
```diff
@@ -0,0 +1,180 @@
+package tailscale_test
```
Comment (oxtoacart) [inline v2/webhooks_test.go:1]: Tests are almost verbatim the same as V1.

File: v2/webhooks_test.go [hunk 6]
```diff
@@ -0,0 +1,180 @@
...
+		ProviderType:     req.ProviderType,
+		CreatorLoginName: "pretend@example.com",
+		Created:          time.Date(2022, 2, 10, 11, 50, 23, 0, time.UTC),
+		LastModified:     time.Date(2022, 2, 10, 11, 50, 23, 0, time.UTC),
+		Subscriptions:    req.Subscriptions,
+		Secret:           &expectedSecret,
+	}
+	server.ResponseBody = expectedWebhook
+
+	webhook, err := client.Webhooks().Create(context.Background(), req)
```
Comment (oxtoacart) [inline v2/webhooks_test.go:38]: Tests were updated to reflect new Resource and corresponding method names.

File: v2/client.go [hunk 7]
```diff
@@ -139,18 +146,29 @@ func requestContentType(ct string) requestOption {
 	}
 }
 
-// buildURL builds a url to /api/v2/... using the given pathElements. It
-// url escapes each path element, so the caller doesn't need to worry about
-// that.
+// buildURL builds a url to /api/v2/... using the given pathElements.
+// It url escapes each path element, so the caller doesn't need to worry about that.
 func (c *Client) buildURL(pathElements ...any) *url.URL {
```
Comment (mpminardi) [inline v2/client.go:151]: Note for ourselves that can be addressed outside of this PR: might be good to have some basic regression unit tests around this method (namely for our expectations around path escaping)

File: v2/client.go [hunk 8]
```diff
@@ -139,18 +146,29 @@ func requestContentType(ct string) requestOption {
...
 	elem := make([]string, 1, len(pathElements)+1)
 	elem[0] = "/api/v2"
 	for _, pathElement := range pathElements {
-		elem = append(elem, fmt.Sprint(pathElement))
+		elem = append(elem, url.PathEscape(fmt.Sprint(pathElement)))
```
Comment (mpminardi) [inline v2/client.go:155]: Another note for future us: will likely want to use [url.QueryEscape](https://pkg.go.dev/net/url#QueryEscape) for any query params we allow interacting with on endpoints that accept them

**Review by mpminardi — APPROVED** [review]
LGTM! Left some notes for future us but nothing blocking

**Pull request merged**